pagefaultgames/pokerogue
# [Bug] Starter UI challenge form fix

## What are the changes?
There's currently a bug with the starter select UI as discussed [here](https://discord.com/channels/1125469663833370665/1269144093938815098)

There's a few underlying bugs that needed to be fixed to fix this bug, so this PR addresses all of them

## Why am I doing these changes?
See above

## What did change?
1. Firstly, the starter UI update from #2830 changed some rules around challenge validity. While the changed files didn't make a different to the logic, they were redundant, so it has been removed.
2. Regarding the bug above from the discord channel, it partly came about because a lot of placed within the starter select screen use a function to get the default dex properties, which is different to the _currently_ chosen dex properties. I added a new function to generate dex properties based on the starter preferences of a pokemon. The other part was that some forms weren't getting checked for challenge validity (i.e. tauros fire form not being checked, so tauros and oricorio would not show on some challenges)
3. Due to the above function, I've made a change where different forms can be selected for challenges, and the logic applies to them. If you are not on a valid form, it will let you know it's not a valid pokemon for the challenge like usual
4. In live at the moment, when going through different pokemon on the starter select screen, a) the icon does not stay if you toggle shiny status and b) if you change variants, the pokemon icon isn't guaranteed to stay either. This has been fixed mainly with the help of using the above function instead of using default props
5. The persistent information for shiny status was not being kept, so this PR adds that functionality. 
6. Currently when choosing a pokemon, the variant star will be at max on the left panel when you move in to update the pokemon's details, regardless of whether you previously chose to have it shiny/an earlier tier of variant. This has now been updated using 
7. Party icons don't update with form/shiny/variant etc. This is now fixed with a new function that updates a party icon to use the current dex properties
8. The behaviour on the starter screen has changed. Previously when choosing a starter, it would choose the highest rank of variant that you have. Now if you don't have a selected choice, it chooses the base non-shiny version of the pokemon for you until you make changes and have them save
9. When reloading, the state of your favourites is not saved (i.e. shiny, form etc) as mentioned in the comment [here](https://github.com/pagefaultgames/pokerogue/pull/3373#pullrequestreview-2221703882). This has now been fixed by adding a section to the updateStarters method that updates the icons of all valid starters when you start a new game. I don't know how performance intensive it is, but it only happens once per new game at the beginning, so hopefully not too bad. See video below on it working

### Screenshots/Videos

This is an example of the behaviour in live. You can see that as I go through pokemon, the shiny and variant status doesn't necessarily stay the same as what I set. Also while I don't point it out, the shiny icon is always the max (in my case, tier 3 shiny):

https://github.com/user-attachments/assets/58b787d8-f0f4-46ee-982d-2368e4597ebf

This is my PR. Apologies for the breakpoint being hit, I hadn't removed it from testing :). As you can see, the shiny and variant status stay the same as what you set, and the shiny icon is the correct variant:

https://github.com/user-attachments/assets/de875d81-2df9-4d33-a3b1-1524ef1c7e7f

Here's the discord bug in beta. As you can see, tauros isn't shown in the fire challenge, nor is oricorio in the electric challenge. Also in the fire challenge, you can change oricorio's form to a non-fire one and it be valid:

https://github.com/user-attachments/assets/36a595c8-8ebc-4954-b657-a0171e7dc36d

Here's my PR fix. Tauros now shows up for a fire challenge, but is only valid to select when you've got the fire form chosen. This is the same for oricorio:

https://github.com/user-attachments/assets/86a0de9f-2bb0-44a7-9f9e-f3690d4c8a3e

Lastly, this is the party icons in live at the moment when you go to change shiny/variant/form:

https://github.com/user-attachments/assets/d84fd77e-2880-4b38-8f61-58bf96d75d88

Finally, this is the same thing in this PR:

https://github.com/user-attachments/assets/0170e354-fcb0-431a-bba7-30a46e988f6d

This is what happens when changing forms/shiny/variants of pokemon and then reloading the page with this PR:

https://github.com/user-attachments/assets/ead9f8c6-4360-4165-978e-8d775798e618

(sorry the video is a little blurry, I had to compress the video to make it fit here) As you can see, all forms/shiny/variant data is now saved when you reload, so all should be good

## How to test the changes?
Download this PR and have a play around with the starter ui. It helps if also have the live/beta version side by side to play around with it and see what's different

## Checklist
